### PR TITLE
ADBDEV-3536 Fix extension creation when extension scripts drop anything

### PR DESCRIPTION
**Pull request by dnskvlnk**
gptkh installation on gpdb5 fails with error:
```
adb=# create extension if not exists gptkh; ERROR: cannot drop function local_tables(text,text,text,integer,text,text) because extension gptkh requires it HINT: You can drop extension gptkh instead.
```
The error is caused by an attempt to delete local_tables function. In gptkh 0.8 there was no attempt to drop this function within the installation script, but in newer version we try to drop some newly created functions, which causes the error.

There’s a difference in extension dependency control between gpdb 5 and gpdb 6. While the 5th forbids dropping extension’s dependencies, the 6th allows to drop dependencies to the extension itself.

Backported test doesn't fully match the test that's on 6x because 5x treats some objects differently. For example, it allows extensions dropping existing views which is forbidden in 5x. However, extensions are not allowed to drop existing functions

The backported commits are as follows:

- https://github.com/arenadata/gpdb/commit/b313cdf2e598fe05ce72a8882493d112f4c44c72 This commit resets extension related variables on segments

- https://github.com/arenadata/gpdb/commit/313786a74155cf04e4539f4d613a6993aab199d8 If the other object is the extension currently being created/altered, ignore this dependency and continue with the deletion.  This allows dropping of an extension's objects within the extension's scripts, as well as corner cases such as dropping a transient object created within such a script.

- https://github.com/arenadata/gpdb/commit/ad20725bf24ea6a85213e6e0379dbd1e665719a0 This patch fixes problem with incorrect handling of `search_path` GUC at QE: `create/alter extension` process sets up this GUC to the newer state, but GUC is not reset after this process finishes - as a result there will broken version of GUC inside transaction. The patch fixes this problem by saving the state of `GUCNestlevel` at QE variable to enable correct state machine work (of `create/alter extension` process) executed on QE.

- https://github.com/arenadata/gpdb/commit/3769eac6d28cfd59b808ff59bfafe264b68649dd The implementation of UPDATE EXTENSION is very similar to CREATE EXTENSION.
First, on QD, we update the pg_extension row to the new version, and update extension's dependency entries according to the new requires introduced by the new version specified control file.
Second, we dispatch the UPDATE EXTENSION statement to QEs to update pg_extension row and extension's dependency entries on QEs, and the global status like creating_extension, CurrentExtensionObject of QEs will be set.
Finally, we execute the update script file on QD. because of the correct set of creating_extension, CurrentExtensionObject on QEs, the dependency between object and extension will be handled correctly, so the CREATE/DROP object statement will be executed successfully.

Patch notes:

1. `test_extension.sql` from the patch differs from the 6x version in the following cases:

    - gpdb5 doesn't support `if not exists` expression, so the `test_ext_cine` test is irrelevant
    - `gp_inject_fault` test was dropped because of different internal organization of this extension
    - `btree_gin` extension doesn't exist in gpdb5 and was replaced by test_ext_cau which causes little differences across the test

2. Signature of ApplyExtensionUpdates was changed 
    This function is supposed to accept AlterExtensionStmt variable in 5x because it's used in ExecAlterExtensionStmt only. But in 6x it's used in CreateExtension as well and doesn't accept any statement at all. The redundant argument was dropped in 6c2e734

3. Add test case for extension dependent function
    Test if we can drop a function which is owned by an extension. In gpdb6 dependency control in handled by DEPENDENCY_INTERNAL_AUTO which doesn't exist in gpdb5, because of this we have to stick to the old dependecy control method. But due to the changes made by this patch, the old contol system broke which led to the situation when an extension couldn't drop its own object during creation or update. This problem was fixed and this test is supposed to check that 

4. Backport guc handling for all gucs
    guc handling was backported from 6x in https://github.com/arenadata/gpdb/commit/ee8daacb503023f9f756922a74fdd2cf8050a736, but GUC_ACTION_SAVE
parameter of set_config_option was lost which led to incorrect guc
handling within transactions. This commit restores original parameter

### CODE REVIEW

**Comment by HustonMmmavr** [comment]
Seems the logic with cleanup global variables (`ResetExtensionCreatingGlobalVarsOnQE`) inside  transactions (in case of abort) should be ported from 6.x branch (there are two places as I see `AbortTransaction` + `AbortOutOfAnyTransaction`)

**Comment by ivanievlev** [comment]
I'm on vacation until Apr 11, 2023.

**Comment by dnskvlnk** [comment]
> Seems the logic with cleanup global variables (`ResetExtensionCreatingGlobalVarsOnQE`) inside transactions (in case of abort) should be ported from 6.x branch (there are two places as I see `AbortTransaction` + `AbortOutOfAnyTransaction`)

Added a proper cleanup in these functions

**Comment by HustonMmmavr** [comment]
There are failed tests at CI, so research (and fix) of problem which leads to this behavior is needed. And it's worth to add tests for current patch, I suppose `src/test/modules/extension` is a suitable place for this (like it's done at 6/7.x)

**Comment by InnerLife0** [comment]
If we backport something, we should add a links to backported commits to PR and commit description. Can you add this, please?

**Comment by dnskvlnk** [comment]
Added links to commits and their descriptions

**Comment by dnskvlnk** [comment]
Done

File: src/backend/catalog/dependency.c [hunk 1]
```diff
@@ -641,7 +641,7 @@ recursiveDeletion(const ObjectAddress *object,
 				 * just ereport here, rather than proceeding, since no other
 				 * dependencies are likely to be interesting.)
 				 */
-				if (callingObject == NULL)
+				if (!creating_extension && callingObject == NULL)
```
Comment (InnerLife0) [inline src/backend/catalog/dependency.c]: Can you say why do we have differences with 6x code in processing of DEPENDENCY_INTERNAL and DEPENDENCY_EXTENSION?

File: src/backend/commands/extension.c [hunk 2]
```diff
@@ -765,13 +768,27 @@ execute_sql_string(const char *sql, const char *filename)
 	CommandCounterIncrement();
 }
 
+static void
+set_end_state(Node *stmt)
```
Comment (InnerLife0) [inline src/backend/commands/extension.c:773]: Additionally to this function, original patch contaned is_begin_state() and a big assertion. Why did we ignore it?

File: src/backend/commands/extension.c [hunk 3]
```diff
@@ -2856,7 +2884,7 @@ ExecAlterExtensionStmt(AlterExtensionStmt *stmt)
  * scripts don't break if newer versions have different control parameters.
  */
 static void
-ApplyExtensionUpdates(AlterExtensionStmt *stmt, Oid extensionOid,
+ApplyExtensionUpdates(Oid extensionOid,
```
Comment (InnerLife0) [inline src/backend/commands/extension.c:2922]: This function has AssertImply() introduced by one of patches. Why did we ignore porting of it?

File: src/backend/commands/extension.c [hunk 4]
```diff
@@ -1238,8 +1253,9 @@ CreateExtension(CreateExtensionStmt *stmt)
 				creating_extension = true;
```
Comment (InnerLife0) [inline src/backend/commands/extension.c]: The latest patch removed this line for a reason. Do we need to remove it too?

Please take a look to latest patch, there are [few(1)](https://github.com/arenadata/gpdb/commit/ad20725bf24ea6a85213e6e0379dbd1e665719a0#diff-f8bf486566de0369065b520e001b0434577000d53dacd135f733931adabc0883L1306) [more(2)](https://github.com/arenadata/gpdb/commit/ad20725bf24ea6a85213e6e0379dbd1e665719a0#diff-f8bf486566de0369065b520e001b0434577000d53dacd135f733931adabc0883L1561) code lines you ignored. Write something if it was for a reason.

File: src/backend/commands/extension.c [hunk 5]
```diff
@@ -100,7 +103,7 @@ static List *find_update_path(List *evi_list,
 static void get_available_versions_for_extension(ExtensionControlFile *pcontrol,
 									 Tuplestorestate *tupstore,
 									 TupleDesc tupdesc);
-static void ApplyExtensionUpdates(AlterExtensionStmt *stmt, Oid extensionOid,
+static void ApplyExtensionUpdates(Oid extensionOid,
```
Comment (InnerLife0) [inline src/backend/commands/extension.c:107]: This change seems to be ported from another patch. Do we need this?

File: src/backend/catalog/dependency.c (hunk 1 above)
Comment (dnskvlnk) [reply src/backend/catalog/dependency.c]: This block of code makes sure that no dependent object is dropped apart from `drop extension`. In gpdb 6 this is handled by `DEPENDENCY_INTERNAL_AUTO` which doesn't exist in gpdb 5, hence we can move it there. `creating_extension` flag is checked to make sure that `drop extension` works correctly, otherwise it will fail.

File: src/backend/commands/extension.c (hunk 3 above)
Comment (dnskvlnk) [reply src/backend/commands/extension.c:2922]: Added to the patch

File: src/backend/commands/extension.c (hunk 4 above)
Comment (dnskvlnk) [reply src/backend/commands/extension.c]: This flag and line 1306 are connected. If you keep `creating_extension = true;` and simplify the check to `if (creating_extension)`, you'll get an error that "nested CREATE EXTENSION is not supported". Modifying both lines makes the patch work correctly, that's fixed. As for the second link, I don't have 'else' block in my patch, please, specify what you meant.

File: src/backend/commands/extension.c (hunk 5 above)
Comment (dnskvlnk) [reply src/backend/commands/extension.c:107]: `stmt` param is out of place here, so I removed it. Should I revert this change?

File: src/backend/commands/extension.c (hunk 2 above)
Comment (dnskvlnk) [reply src/backend/commands/extension.c:773]: Added to the patch

File: src/test/modules/test_extensions/expected/test_extensions.out [hunk 6]
```diff
@@ -73,12 +80,9 @@ SELECT point(0,0) <<@@ polygon(circle(point(0,0),1));
...
- function ext_cor_func()
- operator <<@@(point,polygon)
- type test_ext_type
- view ext_cor_view
-(4 rows)
+                       List of installed extensions
+     Name     | Version | Schema |              Description               
+--------------+---------+--------+----------------------------------------
+ test_ext_cor | 1.0     | public | Test extension using CREATE OR REPLACE
+(1 row)
```
Comment (HustonMmmavr) [inline src/test/modules/test_extensions/expected/test_extensions.out:83]: seems not all test cases exists in this file (also about changing the `search_path` - seems it was missed inside `extension.c`)

File: src/backend/commands/extension.c (hunk 5 above)
Comment (InnerLife0) [reply src/backend/commands/extension.c:107]: I think we should do a refactoring in another patch.

File: src/test/modules/test_extensions/expected/test_extensions.out (hunk 6 above)
Comment (HustonMmmavr) [reply src/test/modules/test_extensions/expected/test_extensions.out:83]: Sorry, `search_path` doesn't missed in `extension.c`. But test cases should be adopted for `5.x`

File: src/backend/catalog/dependency.c (hunk 1 above)
Comment (InnerLife0) [reply src/backend/catalog/dependency.c]: Let's say I want to debug this. Do you have a scenario which will cause an error without `!creating_extension &&`?

File: src/backend/commands/extension.c (hunk 4 above)
Comment (InnerLife0) [reply src/backend/commands/extension.c]: As for 'else' block - one of 6x patches removed it. 5x still has it. Why?
As for the other part, I see it was fixed.

File: src/test/modules/test_extensions/expected/test_extensions.out (hunk 6 above)
Comment (InnerLife0) [reply src/test/modules/test_extensions/expected/test_extensions.out:83]: True. Behavior should be the same. 
I think we can adopt whole test file as well.

Comment (dnskvlnk) [reply src/test/modules/test_extensions/expected/test_extensions.out:83]: Test file was adopted

**Comment by HustonMmmavr** [comment]
To enable tests - seems that [`src/test/modules/Makefile`](https://github.com/arenadata/gpdb/blob/ADBDEV-3536/src/test/modules/Makefile#L10) should be filled like [`contrib/Makefile`](https://github.com/arenadata/gpdb/blob/ADBDEV-3536/contrib/Makefile#L60-L65) and upper-level [GNUMakefile installcheck-world target](https://github.com/arenadata/gpdb/blob/ADBDEV-3536/GNUmakefile.in#L165) should be extended with extension checks

File: src/backend/catalog/dependency.c (hunk 1 above)
Comment (dnskvlnk) [reply src/backend/catalog/dependency.c]: Any that tries to drop dependent object

https://github.com/arenadata/gpdb/blob/4e2f4284c996891977633dc559d8a4d01db236fb/src/test/modules/test_extensions/expected/test_extensions.out#L40-L42

File: src/backend/commands/extension.c (hunk 4 above)
Comment (dnskvlnk) [reply src/backend/commands/extension.c]: Removed redundant code

File: src/backend/commands/extension.c (hunk 5 above)
Comment (dnskvlnk) [reply src/backend/commands/extension.c:107]: Fixed

**Comment by InnerLife0** [comment]
I found some warnings during compilation. Can you please check?
```
extension.c: In function ‘CreateExtension’:
extension.c:1586:25: warning: passing argument 1 of ‘ApplyExtensionUpdates’ from incompatible pointer type [-Wincompatible-pointer-types]
 1586 |   ApplyExtensionUpdates(stmt, extensionOid, pcontrol,
      |                         ^~~~
      |                         |
      |                         CreateExtensionStmt * {aka struct CreateExtensionStmt *}
extension.c:106:55: note: expected ‘AlterExtensionStmt *’ {aka ‘struct AlterExtensionStmt *’} but argument is of type ‘CreateExtensionStmt *’ {aka ‘struct CreateExtensionStmt *’}
  106 | static void ApplyExtensionUpdates(AlterExtensionStmt *stmt, Oid extensionOid,
      |                                   ~~~~~~~~~~~~~~~~~~~~^~~~
extension.c: In function ‘ApplyExtensionUpdates’:
extension.c:3063:40: warning: implicit declaration of function ‘makeDefElem’ [-Wimplicit-function-declaration]
 3063 |    update_stmt->options = lappend(NIL, makeDefElem("new_version", (Node*)makeString(versionName)));
      |                                        ^~~~~~~~~~~
extension.c:3063:40: warning: passing argument 2 of ‘lappend’ makes pointer from integer without a cast [-Wint-conversion]
 3063 |    update_stmt->options = lappend(NIL, makeDefElem("new_version", (Node*)makeString(versionName)));
      |                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                        |
      |                                        int
In file included from ../../../src/include/access/tupdesc.h:19,
                 from ../../../src/include/access/formatter.h:19,
                 from ../../../src/include/access/relscan.h:17,
                 from ../../../src/include/access/genam.h:17,
                 from extension.c:31:
../../../src/include/nodes/pg_list.h:207:14: note: expected ‘void *’ but argument is of type ‘int’
  207 | extern List *lappend(List *list, void *datum);
      |              ^~~~~~~
xact.c: In function ‘AbortTransaction’:
xact.c:3197:3: warning: implicit declaration of function ‘ResetExtensionCreatingGlobalVarsOnQE’ [-Wimplicit-function-declaration]
 3197 |   ResetExtensionCreatingGlobalVarsOnQE();
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: src/test/modules/test_extensions/sql/test_extensions.sql [hunk 7]
```diff
@@ -70,118 +70,3 @@ CREATE EXTENSION test_ext_cine;
 \dx+ test_ext_cine
 ALTER EXTENSION test_ext_cine UPDATE TO '1.1';
 \dx+ test_ext_cine
-
-
-
---
--- Test cases from Issue: https://github.com/greenplum-db/gpdb/issues/6716
---
-drop extension if exists gp_inject_fault;
```
Comment (HustonMmmavr) [inline src/test/modules/test_extensions/sql/test_extensions.sql]: Not all of these test cases should be removed. I suppose tests which uses default extensions (from `contrib`) may be removed, because they don't have an `.control` file at this version of gpdb (postgres)
But test cases with extensions defined inside modules directory (especially, which checks changing `search_path`) should be adopted

**Comment by dnskvlnk** [comment]
@HustonMmmavr @InnerLife0 ready for review

File: src/backend/commands/extension.c [hunk 8]
```diff
@@ -862,7 +863,7 @@ execute_extension_script(Node *stmt,
 
 	(void) set_config_option("search_path", pathbuf.data,
 							 PGC_USERSET, PGC_S_SESSION,
-							 true, true);
+							 GUC_ACTION_SAVE, true);
```
Comment (InnerLife0) [inline src/backend/commands/extension.c:866]: What was the reason of adding GUC_ACTION_SAVE here?

File: src/backend/commands/extension.c [hunk 9]
```diff
@@ -2918,7 +2919,7 @@ ExecAlterExtensionStmt(AlterExtensionStmt *stmt)
  * scripts don't break if newer versions have different control parameters.
  */
 static void
-ApplyExtensionUpdates(AlterExtensionStmt *stmt, Oid extensionOid,
+ApplyExtensionUpdates(Oid extensionOid,
```
Comment (InnerLife0) [inline src/backend/commands/extension.c:2922]: And what was the reason of refactoring this in current patch?

File: src/backend/commands/extension.c (hunk 8 above)
Comment (dnskvlnk) [reply src/backend/commands/extension.c:866]: This flag is a direct port of 6x's logic. gpdb5 doesn't save nested gucs correctly without this flag. Moreover, it doesn't have any tests checking its correctness. It was fixed and test suit was added

File: src/backend/commands/extension.c (hunk 9 above)
Comment (dnskvlnk) [reply src/backend/commands/extension.c:2922]: This function is supposed to accept `AlterExtensionStmt` variable in 5x because it's used in `ExecAlterExtensionStmt` only. But in 6x it's used in `CreateExtension` as well and doesn't accept any statement at all. `AlterExtensionStmt` and `CreateExtensionStmt` don't have any base type so there's nothing to cast to.

File: src/backend/commands/extension.c (hunk 8 above)
Comment (HustonMmmavr) [reply src/backend/commands/extension.c:866]: There was an incorrectly passed argument to the `set_config_option` call (it expects an `GucAction` enum), which leads to an incorrect handling GUC inside transaction (and seems not only transaction), for example [this test case](
https://github.com/arenadata/gpdb/pull/493/files/0510b5f831e198a5271f41c29c126cedf6a6ae5c..a475214bda539a1cfeeb5b5d35fd6efcb3c88c8e#diff-bdeea723c8fc7705b36694bcbd9282603aebcbecbfbc113499dd7f84f5693e63R105-R121) won't work withoout this change:

```sql
create schema differentschema;
begin;
set search_path=differentschema;
show search_path;
   search_path   
-----------------
 differentschema
(1 row)

create extension test_ext_cau with schema pg_catalog;
show search_path;
   search_path   
-----------------
 differentschema # it returns now , but with an incorrect `true` it would be pg_catalog;
(1 row)

end;
```

Comment (InnerLife0) [reply src/backend/commands/extension.c:866]: 1) Why this placed in commit named "Fix compilation warnings"?
2) Where is any explanation (PR message, commit message) of this?

File: src/backend/commands/extension.c (hunk 9 above)
Comment (InnerLife0) [reply src/backend/commands/extension.c:2922]: > `AlterExtensionStmt` and `CreateExtensionStmt` don't have any base type so there's nothing to cast to.

They both Nodes, at least.

I don't see any single comment (in the right place) about importance of refactoring.

**Comment by HustonMmmavr** [comment]
Also there are a broken pipeline at regression tests with postgres

File: src/backend/catalog/dependency.c (hunk 1 above)
Comment (InnerLife0) [reply src/backend/catalog/dependency.c]: I got the same error for the given snippet even without using of `creating_extension` flag.
Do you have a scenario which shows unwanted behavior without this line?

**Comment by InnerLife0** [comment]
True, please check the tests.

Also, I see some differences between 6x and 5x versions of test_extensions.sql. Can you briefly describe the difference and why our tests changed?

File: src/test/modules/test_extensions/sql/test_extensions.sql (hunk 7 above)
Comment (dnskvlnk) [reply src/test/modules/test_extensions/sql/test_extensions.sql]: Tests were reworked

File: src/backend/commands/extension.c (hunk 9 above)
Comment (dnskvlnk) [reply src/backend/commands/extension.c:2922]: Actually, why do we consider this as refactoring? `ApplyExtensionUpdates` doesn't have this arg in 6x, so this is a straight port of this code base. Also, if we want to keep this argument we need to change this line anyway. Why not just remove it to avoid any confusions? I think this diff can be really confusing considering that this argument is not used in the function at all, but we still keep it

Comment (InnerLife0) [reply src/backend/commands/extension.c:2922]: You already explained your point of view. I mentioned, that I don't see a comment in the right place.
Explain its importance in commit and PR comments and give a link to 6x commit as reference. That will be OK.

**Comment by dnskvlnk** [comment]
@HustonMmmavr there's no need to remove `drop function test_func1` statement because this function stays after the extension is updated to 1.1, this behaviour is consistent with 6x

**Comment by dnskvlnk** [comment]
`test_extension.sql` from the patch differs from the 6x version in the following cases:

- gpdb5 doesn't support `if not exists` expression, so the `test_ext_cine` test is irrelevant
- `gp_inject_fault` test was dropped because of different internal organization of this extension
- `btree_gin` extension doesn't exist in gpdb5 and was replaced by test_ext_cau which causes little differences across the test

File: src/backend/catalog/dependency.c (hunk 1 above)
Comment (dnskvlnk) [reply src/backend/catalog/dependency.c]: Test case was added

Comment (dnskvlnk) [reply src/backend/catalog/dependency.c]: https://github.com/arenadata/gpdb/blob/599e1286a8caf54431ae39a40bf2f4f3632cb52b/src/test/modules/test_extensions/expected/test_extensions.out#L51-L53

File: src/backend/commands/extension.c (hunk 9 above)
Comment (dnskvlnk) [reply src/backend/commands/extension.c:2922]: Description added

File: src/backend/commands/extension.c (hunk 8 above)
Comment (dnskvlnk) [reply src/backend/commands/extension.c:865]: Description added

Comment (dnskvlnk) [reply src/backend/commands/extension.c:866]: Description added

File: src/backend/catalog/dependency.c (hunk 1 above)
Comment (dnskvlnk) [reply src/backend/catalog/dependency.c]: The flag was removed, test case was added

**Review by InnerLife0 — DISMISSED** [review]
Approved. You need to squash all commits and update commit message with the latest revision after second approval.

**Review by HustonMmmavr — DISMISSED** [review]

**Review by InnerLife0 — APPROVED** [review]

**Review by HustonMmmavr — APPROVED** [review]
